openssl/src/hkdf.rs

```diff
@@ -0,0 +1,109 @@
+use std::ptr;
+
+use libc::c_int;
+
+use error::ErrorStack;
+use ffi;
+use hash::MessageDigest;
+use {cvt, cvt_p};
+
+pub struct Hkdf(*mut ffi::EVP_PKEY_CTX);
+
+pub enum Mode {
+    ExtractAndExpand,
+    ExtractOnly,
+    ExpandOnly,
+}
+
+unsafe impl Send for Hkdf {}
+
+impl Hkdf {
+    pub fn new(hash: MessageDigest) -> Result<Self, ErrorStack> {
+        unsafe {
+            let ret = cvt_p(ffi::EVP_PKEY_CTX_new_id(
+                ffi::EVP_PKEY_HKDF,
+                ptr::null_mut(),
+            ))
+            .map(Hkdf)
+            .and_then(|ctx| cvt(ffi::EVP_PKEY_derive_init(ctx.0)).map(|_| ctx))?;
+
+            cvt(ffi::EVP_PKEY_CTX_set_hkdf_md(ret.0, hash.as_ptr()))?;
+
+            Ok(ret)
+        }
+    }
+
+    pub fn set_mode(self, mode: Mode) -> Result<Self, ErrorStack> {
```

Comment:
I am not sure if by-value builds are a good choice here as existing builds like `X509Builder` use the by-mut-ref pattern instead. But more importantly, there is not way to recover the value if there is recoverable error (e.g. an unsupported and optional operation) while using a by-value fallible builder which does not return itself on the error arm as well (like [`try_send`](https://doc.rust-lang.org/stable/std/sync/mpsc/struct.SyncSender.html#method.try_send) from the standard library).

Reply:
Looking at #1234, it seems questionable whether a builder pattern is applicable at all as these parameters are not optional or have sensible defaults. This could be encoded via a type state, but that feels like unhelpful ceremony compared to a free function.